rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,227 @@
+
+We will add capability to use optional annotations to better control transformation of Kubernetes services into Teleport apps.
+By default we will be exposing services as `tcp` apps, since it's the most general type, but annotation or exposed port's `appProtocol` field
+can be used to specify which protocol to use for the app's URI. 
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels:
+        env: staging
+    - types: ["app"]
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
+```
+
+`kubernetes_cluster` field in the config will be translated into `teleport.dev/kubernetes-cluster` label on the Teleport app resource, so
+later it's be easier to target it in the corresponding app service.
+
+#### Annotations
+
+Kubernetes annotations will allow users to fine tune transformation of services into Teleport apps. They will override default behaviour, 
+but they are not required for import of services - by default services without any annotations will also be imported.
+
+Annotation `teleport.dev/discovery-type` controls what Teleport entity will be created from this service, for now we will only support value 
```

Comment:
I'd keep this one out of scope for now.

Reply:
Why do we need this annotation if the resource created within Teleport is controlled by the matcher's "type" field?

Reply:
That's so the matcher can filter out service it should and shouldn't pick up. If we have service that has annoations `teleport.dev/discovery-type: "db"`, then matcher configured for "app" will skip it.
But come to think of it, we probably can use annotation for protocol for this, if it would say "postgresql", "mysql" etc we know it's database type 🤔

Reply:
I'm still confused by this tbh, imagine you have the following discovery configuration

```yaml
kubernetes:
- types: ["app"]
  labels:
    a: b
```

but it matches a service with label `discovery-type: db`. What will it create, app or db?

That's why I'm proposing to get rid of this annotation and have the discovery configuration govern what kind of resources it creates for simplicity.

Reply:
Type is part of matching, idea is it says what we are locking for, not what we are creating (it's like a special cased label). All services by default are just assumed to be of type 'app' and annotation can override it. So service with label `discovery-type: db` will not be matched and nothing will be imported from this service. If user adds to config 
```yaml
kubernetes:
- types: ["db"]
  labels:
    a: b
```
Then db will be created from the service, because user says "This service is database" and discovery service is configured to look for databases. Users can also easily exclude any service from being matched fx by setting annotation `teleport.dev/discovery-type: none`

Generally, once DB is also supported, when we're looking at a service we don't know what it is, if it's an app or a db, so we need some hint to understand to what Teleport entity map it. But as I mentioned above, we probably can use protocol annotation, because `teleport.dev/protocol: postgres` will definitely mean that this is a database.
To me the path with dedicated `types` field and annotation look more clear and future-proof, but I'm not so sure now, since it didn't look clear to you. @tigrato what are your thoughts?